src/librustc_typeck/check/method/mod.rs

```diff
@@ -340,6 +345,42 @@ pub fn report_error<'a, 'tcx>(fcx: &FnCtxt<'a, 'tcx>,
 
                 report_candidates(fcx, span, method_name, static_sources);
             }
+
+            let candidates = all_traits(fcx.ccx)
+                .filter(|did| {
+                    debug!("report_error: looking at {:?}, {}.",
+                           did, ty::item_path_str(fcx.tcx(), *did));
+                    ty::trait_items(fcx.tcx(), *did)
+                        .iter()
+                        .any(|item| {
+                            debug!("report_error:    item: {:?}", *item);
+                            match *item {
+                                ty::MethodTraitItem(ref meth) => meth.name == method_name,
+                                ty::TypeTraitItem(_) => false
+                            }
+                        })
+                })
+                .collect::<Vec<_>>();
+
+            if candidates.len() > 0 {
+                span_help!(fcx.sess(), span,
+                           "methods from traits can only be called if the trait is implemented \
+                            and in scope; the following trait{s} define{inv_s} a method `{name}`:",
+                           s = if candidates.len() == 1 {""} else {"s"},
+                           inv_s = if candidates.len() == 1 {"s"} else {""},
+                           name = method_ustring);
+
+                for (i, &did) in candidates.iter().enumerate() {
+                    // provide a good-as-possible span; the span of
+                    // the trait if it is local, or the span of the
+                    // method call itself if not
+                    let trait_span = fcx.tcx().map.def_id_span(did, span);
```

Comment:
I think it'd be good to cross-reference this against the list of traits that is imported. Here are some tips to gather more info:

If we want to decide if the method is maybe implemented for this type, it requires iterating down the steps and invoking one of the "evaluate_obligation" methods on a trait context. The tricky part is we also have to deal with autoref and the method self-type and so forth. It's kind of a pain but certainly possible in the name of a great error message. I can help you with that but let's discuss on IRC.